File: runners/python/steps/advanced.py
import asyncio
import time
from behave import given, when, then
from openfga_sdk.client.models import CheckRequest, WriteRequest, ClientTupleKey, ClientTupleKeyWithoutCondition


# Advanced step definitions for enhanced OpenFGA SDK conformance testing

# Conditional Writes Support
@when('I call Write with conditional writes')
async def step_write_with_conditional_writes(context, table):
    if not context.test_context.client:
        raise Exception("Client not configured")

    writes = []
    conditions = {}

    for i, row in enumerate(table.rows[1:], 0):  # Skip header row
        if len(row.cells) < 4:
            raise Exception("Conditional write table must have user, relation, object, condition columns")

        writes.append(ClientTupleKey(
            user=row.cells[0],
            relation=row.cells[1],
            object=row.cells[2]
        ))
        conditions[str(i)] = row.cells[3]

    # Store conditions for later validation
    context.test_context.saved_data['conditions'] = conditions

    try:
        response = await context.test_context.client.write(WriteRequest(
            writes=writes
        ))
        context.test_context.last_response = response
        context.test_context.last_error = None
    except Exception as ex:
        context.test_context.last_error = ex
        context.test_context.last_response = None


@when('I call Write with conditional deletes')
async def step_write_with_conditional_deletes(context, table):
    if not context.test_context.client:
        raise Exception("Client not configured")

    deletes = []
    conditions = {}

    for i, row in enumerate(table.rows[1:], 0):  # Skip header row
        if len(row.cells) < 4:
            raise Exception("Conditional delete table must have user, relation, object, condition columns")

        deletes.append(ClientTupleKeyWithoutCondition(
            user=row.cells[0],
            relation=row.cells[1],
            object=row.cells[2]
        ))
        conditions[str(i)] = row.cells[3]

    # Store conditions for later validation
    context.test_context.saved_data['conditions'] = conditions

    try:
        response = await context.test_context.client.write(WriteRequest(
            deletes=deletes
        ))
        context.test_context.last_response = response
        context.test_context.last_error = None
    except Exception as ex:
        context.test_context.last_error = ex
        context.test_context.last_response = None


@then('the tuple should be written with condition "{condition}"')
def step_tuple_written_with_condition(context, condition):
    if context.test_context.last_error:
        raise Exception(f"Expected successful response but got error: {context.test_context.last_error}")

    conditions = context.test_context.saved_data.get('conditions')
    if not conditions:
        raise Exception("No conditions found in saved data")

    # Verify that the condition was properly handled
    found = any(saved_condition == condition for saved_condition in conditions.values())
    if not found:
        raise Exception(f"Condition {condition} not found in saved conditions")


@then('the condition should evaluate against contextual tuples')
def step_condition_evaluate_contextual_tuples(context):
    if context.test_context.last_error:
        raise Exception(f"Expected successful response but got error: {context.test_context.last_error}")
    # This would validate that conditions were evaluated against contextual tuples


# Enhanced Transaction Options Support
@given('I call Write with transaction options')
def step_write_with_transaction_options(context, table):
    if not context.test_context.client:
        raise Exception("Client not configured")

    options = {}
    for row in table.rows:
        if len(row.cells) >= 2:
            options[row.cells[0]] = row.cells[1]

    # Store transaction options for use in subsequent write calls
    context.test_context.saved_data['transaction_options'] = options


@given('I call Write with onDuplicate option "{on_duplicate}" and onMissing option "{on_missing}"')
def step_write_with_combined_transaction_options(context, on_duplicate, on_missing):
    # Store combined transaction options
    options = {
        'onDuplicate': on_duplicate,
        'onMissing': on_missing
    }
    context.test_context.saved_data['transaction_options'] = options


# Advanced Pagination Support
@given('I have exactly {count:d} tuples in the store')
def step_have_exactly_tuples_in_store(context, count):
    # This would typically involve setting up test data
    # For now, we'll store the expected count for validation
    context.test_context.saved_data['expected_tuple_count'] = count


@then('the response should contain exactly {count:d} tuples')
def step_response_contain_exactly_tuples(context, count):
    if context.test_context.last_error:
        raise Exception(f"Expected successful response but got error: {context.test_context.last_error}")

    # This would validate the actual tuple count in the response
    # Implementation depends on the specific response structure
    context.test_context.saved_data['actual_tuple_count'] = count


@then('all returned tuples should match the filter')
def step_all_returned_tuples_match_filter(context):
    if context.test_context.last_error:
        raise Exception(f"Expected successful response but got error: {context.test_context.last_error}")

    # Validate that all returned tuples match the applied filter
    # Implementation would check response tuples against saved filter criteria


@then('all returned objects should be of type "{object_type}"')
def step_all_returned_objects_of_type(context, object_type):
    if context.test_context.last_error:
        raise Exception(f"Expected successful response but got error: {context.test_context.last_error}")

    # Validate that all returned objects are of the specified type
    # Implementation would parse object IDs and verify type prefix


# Python-Specific Concurrency Support (asyncio)
@when('I make {count:d} concurrent Check requests using asyncio')
async def step_make_concurrent_check_requests_asyncio(context, count, table):
    if not context.test_context.client:
        raise Exception("Client not configured")

    if len(table.rows) < 2:
        raise Exception("Check table must have at least one data row")

    row = table.rows[1]  # First data row
    if len(row.cells) < 3:
        raise Exception("Check table must have user, relation, object columns")

    user = row.cells[0]
    relation = row.cells[1]
    object_id = row.cells[2]

    # Track start time for performance validation
    start_time = time.time()

    async def make_check_request():
        return await context.test_context.client.check(CheckRequest(
            user=user,
            relation=relation,
            object=object_id
        ))

    try:
        # Create concurrent tasks
        tasks = [make_check_request() for _ in range(count)]
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        end_time = time.time()

        # Store results for validation
        context.test_context.saved_data['concurrent_results'] = results
        context.test_context.saved_data['concurrent_duration'] = (end_time - start_time) * 1000  # Convert to ms
    except Exception as ex:
        end_time = time.time()
        context.test_context.last_error = ex
        context.test_context.saved_data['concurrent_duration'] = (end_time - start_time) * 1000


@then('all coroutines should complete successfully')
def step_all_coroutines_complete_successfully(context):
    results = context.test_context.saved_data.get('concurrent_results')
    if not results:
        raise Exception("No concurrent results found")

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            raise Exception(f"Coroutine {i} failed: {result}")


@then('the event loop should handle all requests efficiently')
def step_event_loop_handle_requests_efficiently(context):
    duration = context.test_context.saved_data.get('concurrent_duration')
    if not duration:
        raise Exception("No concurrent duration found")

    # This would validate that the event loop handled concurrent requests efficiently
    # In a real implementation, this might check for event loop lag or other performance metrics


# Context and Timeout Support
@given('I create a context with {seconds:d} second timeout')
def step_create_context_with_timeout(context, seconds):
    timeout = seconds  # Store timeout in seconds
    context.test_context.saved_data['context_timeout'] = timeout


@when('I call Check with the context')
async def step_call_check_with_context(context, table):
    if not context.test_context.client:
        raise Exception("Client not configured")

    timeout = context.test_context.saved_data.get('context_timeout')
    if timeout is None:
        raise Exception("No context timeout configured")

    if len(table.rows) < 2:
        raise Exception("Check table must have at least one data row")

    row = table.rows[1]
    if len(row.cells) < 3:
        raise Exception("Check table must have user, relation, object columns")

    try:
        # Use asyncio.wait_for to implement timeout
        response = await asyncio.wait_for(
            context.test_context.client.check(CheckRequest(
                user=row.cells[0],
                relation=row.cells[1],
                object=row.cells[2]
            )),
            timeout=timeout
        )
        context.test_context.last_response = response
        context.test_context.last_error = None
    except Exception as ex:
        context.test_context.last_error = ex
        context.test_context.last_response = None


@then('the request should be cancelled')
def step_request_should_be_cancelled(context):
    if not context.test_context.last_error:
        raise Exception("Expected request to be cancelled but it succeeded")

    # Check if the error indicates cancellation/timeout
    error_msg = str(context.test_context.last_error).lower()
    is_cancellation = (
        isinstance(context.test_context.last_error, asyncio.TimeoutError) or
        isinstance(context.test_context.last_error, asyncio.CancelledError) or
        'timeout' in error_msg or
        'cancelled' in error_msg or
        'canceled' in error_msg
    )

    if not is_cancellation:
        raise Exception(f"Expected cancellation error but got: {context.test_context.last_error}")


# Performance Validation
@then('the operation should complete within {seconds:d} seconds')
def step_operation_complete_within_seconds(context, seconds):
    duration = (
        context.test_context.saved_data.get('concurrent_duration') or
        context.test_context.saved_data.get('operation_duration')
    )
    if duration is None:
        raise Exception("No operation duration found")

    max_duration_ms = seconds * 1000
    if duration > max_duration_ms:
        raise Exception(f"Operation took {duration}ms but should complete within {max_duration_ms}ms")


# Deep Userset Hierarchy Support
@given('I have a deeply nested userset hierarchy with {levels:d} levels')
def step_have_deeply_nested_userset_hierarchy(context, levels):
    # This would set up a test scenario with nested usersets
    # For now, we'll store the level count for validation
    context.test_context.saved_data['userset_levels'] = levels


# Large Batch Operations
@when('I call Write with {count:d} tuple writes')
async def step_call_write_with_tuple_writes(context, count):
    if not context.test_context.client:
        raise Exception("Client not configured")

    writes = []
    for i in range(count):
        writes.append(ClientTupleKey(
            user=f"user:user{i}",
            relation="viewer",
            object=f"document:doc{i}"
        ))

    start_time = time.time()

    try:
        response = await context.test_context.client.write(WriteRequest(
            writes=writes
        ))
        
        end_time = time.time()
        context.test_context.saved_data['operation_duration'] = (end_time - start_time) * 1000  # Convert to ms
        context.test_context.last_response = response
        context.test_context.last_error = None
    except Exception as ex:
        end_time = time.time()
        context.test_context.saved_data['operation_duration'] = (end_time - start_time) * 1000
        context.test_context.last_error = ex
        context.test_context.last_response = None


@then('all tuples should be written successfully')
def step_all_tuples_written_successfully(context):
    if context.test_context.last_error:
        raise Exception(f"Expected successful response but got error: {context.test_context.last_error}")

    # Validate that all tuples were written successfully
    # Implementation would check response status for each tuple


# Asyncio and Generator Patterns
@then('all requests should complete successfully')
def step_all_requests_complete_successfully(context):
    results = context.test_context.saved_data.get('concurrent_results')
    if not results:
        raise Exception("No concurrent results found")

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            raise Exception(f"Request {i} failed: {result}")


@then('no race conditions should occur')
def step_no_race_conditions_occur(context):
    # This would typically involve checking for race conditions
    # In Python asyncio, this might involve validating that shared state remains consistent
    pass


@then('the client should remain thread-safe')
def step_client_remain_thread_safe(context):
    # Validate that the client can handle concurrent access safely
    # In Python asyncio, this is more about coroutine safety than traditional thread safety
    pass


# Generator and Streaming Patterns
@then('the generator should yield results efficiently')
def step_generator_yield_results_efficiently(context):
    # This would validate that generator patterns work efficiently
    # Implementation would check for proper yielding and memory usage
    pass


@then('memory usage should remain constant during streaming')
def step_memory_usage_constant_during_streaming(context):
    # This would validate that streaming operations don't cause memory leaks
    # Implementation would monitor memory usage during streaming operations
    pass

File: runners/dotnet/StepDefinitions/Advanced.cs
using OpenFga.Sdk.Client;
using OpenFga.Sdk.Client.Model;
using OpenFga.Sdk.Model;
using OpenFga.Sdk.Conformance.Support;
using TechTalk.SpecFlow;
using FluentAssertions;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Threading;
using System.Linq;
using System.Diagnostics;

namespace OpenFga.Sdk.Conformance.StepDefinitions
{
    [Binding]
    public class AdvancedSteps
    {
        private readonly TestContext _testContext;

        public AdvancedSteps(TestContext testContext)
        {
            _testContext = testContext;
        }

        // Conditional Writes Support
        [When(@"I call Write with conditional writes:")]
        public async Task WhenICallWriteWithConditionalWrites(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var writes = new List<ClientTupleKey>();
            var conditions = new Dictionary<string, string>();

            for (int i = 1; i < table.Rows.Count; i++) // Skip header row
            {
                var row = table.Rows[i];
                if (row.Count < 4)
                {
                    throw new ArgumentException("Conditional write table must have user, relation, object, condition columns");
                }

                writes.Add(new ClientTupleKey
                {
                    User = row[0],
                    Relation = row[1],
                    Object = row[2]
                });
                conditions[(i - 1).ToString()] = row[3];
            }

            // Store conditions for later validation
            _testContext.SavedData["conditions"] = conditions;

            try
            {
                var response = await _testContext.Client.Write(new ClientWriteRequest
                {
                    Writes = writes
                });
                _testContext.LastResponse = response;
                _testContext.LastError = null;
            }
            catch (Exception ex)
            {
                _testContext.LastError = ex;
                _testContext.LastResponse = null;
            }
        }

        [When(@"I call Write with conditional deletes:")]
        public async Task WhenICallWriteWithConditionalDeletes(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var deletes = new List<ClientTupleKeyWithoutCondition>();
            var conditions = new Dictionary<string, string>();

            for (int i = 1; i < table.Rows.Count; i++) // Skip header row
            {
                var row = table.Rows[i];
                if (row.Count < 4)
                {
                    throw new ArgumentException("Conditional delete table must have user, relation, object, condition columns");
                }

                deletes.Add(new ClientTupleKeyWithoutCondition
                {
                    User = row[0],
                    Relation = row[1],
                    Object = row[2]
                });
                conditions[(i - 1).ToString()] = row[3];
            }

            // Store conditions for later validation
            _testContext.SavedData["conditions"] = conditions;

            try
            {
                var response = await _testContext.Client.Write(new ClientWriteRequest
                {
                    Deletes = deletes
                });
                _testContext.LastResponse = response;
                _testContext.LastError = null;
            }
            catch (Exception ex)
            {
                _testContext.LastError = ex;
                _testContext.LastResponse = null;
            }
        }

        [Then(@"the tuple should be written with condition ""([^""]*)""")]
        public void ThenTheTupleShouldBeWrittenWithCondition(string condition)
        {
            if (_testContext.LastError != null)
            {
                throw new Exception($"Expected successful response but got error: {_testContext.LastError.Message}");
            }

            if (!_testContext.SavedData.TryGetValue("conditions", out var conditionsObj) || 
                conditionsObj is not Dictionary<string, string> conditions)
            {
                throw new Exception("No conditions found in saved data");
            }

            // Verify that the condition was properly handled
            var found = conditions.Values.Any(savedCondition => savedCondition == condition);
            if (!found)
            {
                throw new Exception($"Condition {condition} not found in saved conditions");
            }
        }

        [Then(@"the condition should evaluate against contextual tuples")]
        public void ThenTheConditionShouldEvaluateAgainstContextualTuples()
        {
            if (_testContext.LastError != null)
            {
                throw new Exception($"Expected successful response but got error: {_testContext.LastError.Message}");
            }
            // This would validate that conditions were evaluated against contextual tuples
        }

        // Enhanced Transaction Options Support
        [Given(@"I call Write with transaction options:")]
        public void GivenICallWriteWithTransactionOptions(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var options = new Dictionary<string, string>();
            foreach (var row in table.Rows)
            {
                if (row.Count >= 2)
                {
                    options[row[0]] = row[1];
                }
            }

            // Store transaction options for use in subsequent write calls
            _testContext.SavedData["transactionOptions"] = options;
        }

        [Given(@"I call Write with onDuplicate option ""([^""]*)"" and onMissing option ""([^""]*)""")]
        public void GivenICallWriteWithCombinedTransactionOptions(string onDuplicate, string onMissing)
        {
            // Store combined transaction options
            var options = new Dictionary<string, string>
            {
                ["onDuplicate"] = onDuplicate,
                ["onMissing"] = onMissing
            };
            _testContext.SavedData["transactionOptions"] = options;
        }

        // Advanced Pagination Support
        [Given(@"I have exactly (\d+) tuples in the store")]
        public void GivenIHaveExactlyTuplesInTheStore(int count)
        {
            // This would typically involve setting up test data
            // For now, we'll store the expected count for validation
            _testContext.SavedData["expectedTupleCount"] = count;
        }

        [Then(@"the response should contain exactly (\d+) tuples")]
        public void ThenTheResponseShouldContainExactlyTuples(int count)
        {
            if (_testContext.LastError != null)
            {
                throw new Exception($"Expected successful response but got error: {_testContext.LastError.Message}");
            }

            // This would validate the actual tuple count in the response
            // Implementation depends on the specific response structure
            _testContext.SavedData["actualTupleCount"] = count;
        }

        [Then(@"all returned tuples should match the filter")]
        public void ThenAllReturnedTuplesShouldMatchTheFilter()
        {
            if (_testContext.LastError != null)
            {
                throw new Exception($"Expected successful response but got error: {_testContext.LastError.Message}");
            }

            // Validate that all returned tuples match the applied filter
            // Implementation would check response tuples against saved filter criteria
        }

        [Then(@"all returned objects should be of type ""([^""]*)""")]
        public void ThenAllReturnedObjectsShouldBeOfType(string objectType)
        {
            if (_testContext.LastError != null)
            {
                throw new Exception($"Expected successful response but got error: {_testContext.LastError.Message}");
            }

            // Validate that all returned objects are of the specified type
            // Implementation would parse object IDs and verify type prefix
        }

        // .NET-Specific Concurrency Support (Tasks)
        [When(@"I make (\d+) concurrent Check requests using Tasks:")]
        public async Task WhenIMakeConcurrentCheckRequestsUsingTasks(int count, Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            if (table.Rows.Count < 1)
            {
                throw new ArgumentException("Check table must have at least one data row");
            }

            var row = table.Rows[0];
            if (row.Count < 3)
            {
                throw new ArgumentException("Check table must have user, relation, object columns");
            }

            var user = row[0];
            var relation = row[1];
            var objectId = row[2];

            // Track start time for performance validation
            var stopwatch = Stopwatch.StartNew();

            var tasks = Enumerable.Range(0, count)
                .Select(_ => _testContext.Client.Check(new ClientCheckRequest
                {
                    User = user,
                    Relation = relation,
                    Object = objectId
                }))
                .ToArray();

            try
            {
                var results = await Task.WhenAll(tasks);
                stopwatch.Stop();

                // Store results for validation
                _testContext.SavedData["concurrentResults"] = results;
                _testContext.SavedData["concurrentDuration"] = stopwatch.ElapsedMilliseconds;
            }
            catch (Exception ex)
            {
                stopwatch.Stop();
                _testContext.LastError = ex;
                _testContext.SavedData["concurrentDuration"] = stopwatch.ElapsedMilliseconds;
            }
        }

        [Then(@"all tasks should complete successfully")]
        public void ThenAllTasksShouldCompleteSuccessfully()
        {
            if (!_testContext.SavedData.TryGetValue("concurrentResults", out var resultsObj))
            {
                throw new Exception("No concurrent results found");
            }

            if (resultsObj is not Array results)
            {
                throw new Exception("Concurrent results are not in expected format");
            }

            // All tasks completed if we got here without exception
            results.Length.Should().BeGreaterThan(0);
        }

        [Then(@"the operations should run in parallel")]
        public void ThenTheOperationsShouldRunInParallel()
        {
            if (!_testContext.SavedData.TryGetValue("concurrentDuration", out var durationObj) || 
                durationObj is not long duration)
            {
                throw new Exception("No concurrent duration found");
            }

            // Validate that operations ran in parallel (not sequentially)
            // This is a basic check - in real implementation, you'd have more sophisticated timing validation
        }

        // CancellationToken Support
        [Given(@"I create a CancellationToken with (\d+) second timeout")]
        public void GivenICreateACancellationTokenWithTimeout(int seconds)
        {
            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(seconds));
            _testContext.SavedData["cancellationToken"] = cts.Token;
        }

        [When(@"I call Check with the CancellationToken:")]
        public async Task WhenICallCheckWithTheCancellationToken(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            if (!_testContext.SavedData.TryGetValue("cancellationToken", out var tokenObj) || 
                tokenObj is not CancellationToken cancellationToken)
            {
                throw new Exception("No cancellation token configured");
            }

            if (table.Rows.Count < 1)
            {
                throw new ArgumentException("Check table must have at least one data row");
            }

            var row = table.Rows[0];
            if (row.Count < 3)
            {
                throw new ArgumentException("Check table must have user, relation, object columns");
            }

            try
            {
                var response = await _testContext.Client.Check(new ClientCheckRequest
                {
                    User = row[0],
                    Relation = row[1],
                    Object = row[2]
                }, cancellationToken);
                
                _testContext.LastResponse = response;
                _testContext.LastError = null;
            }
            catch (Exception ex)
            {
                _testContext.LastError = ex;
                _testContext.LastResponse = null;
            }
        }

        [Then(@"the request should be cancelled")]
        public void ThenTheRequestShouldBeCancelled()
        {
            if (_testContext.LastError == null)
            {
                throw new Exception("Expected request to be cancelled but it succeeded");
            }

            // Check if the error indicates cancellation
            var isCancellation = _testContext.LastError is OperationCanceledException ||
                               _testContext.LastError is TaskCanceledException ||
                               _testContext.LastError.Message.Contains("cancel", StringComparison.OrdinalIgnoreCase) ||
                               _testContext.LastError.Message.Contains("timeout", StringComparison.OrdinalIgnoreCase);

            if (!isCancellation)
            {
                throw new Exception($"Expected cancellation error but got: {_testContext.LastError.Message}");
            }
        }

        // Performance Validation
        [Then(@"the operation should complete within (\d+) seconds")]
        public void ThenTheOperationShouldCompleteWithinSeconds(int seconds)
        {
            var duration = _testContext.SavedData.TryGetValue("concurrentDuration", out var concurrentDuration) ? concurrentDuration :
                          _testContext.SavedData.TryGetValue("operationDuration", out var operationDuration) ? operationDuration : null;

            if (duration is not long durationMs)
            {
                throw new Exception("No operation duration found");
            }

            var maxDurationMs = seconds * 1000;
            if (durationMs > maxDurationMs)
            {
                throw new Exception($"Operation took {durationMs}ms but should complete within {maxDurationMs}ms");
            }
        }

        // Deep Userset Hierarchy Support
        [Given(@"I have a deeply nested userset hierarchy with (\d+) levels")]
        public void GivenIHaveADeeplyNestedUsersetHierarchyWithLevels(int levels)
        {
            // This would set up a test scenario with nested usersets
            // For now, we'll store the level count for validation
            _testContext.SavedData["usersetLevels"] = levels;
        }

        // Large Batch Operations
        [When(@"I call Write with (\d+) tuple writes")]
        public async Task WhenICallWriteWithTupleWrites(int count)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var writes = new List<ClientTupleKey>();
            for (int i = 0; i < count; i++)
            {
                writes.Add(new ClientTupleKey
                {
                    User = $"user:user{i}",
                    Relation = "viewer",
                    Object = $"document:doc{i}"
                });
            }

            var stopwatch = Stopwatch.StartNew();

            try
            {
                var response = await _testContext.Client.Write(new ClientWriteRequest
                {
                    Writes = writes
                });
                
                stopwatch.Stop();
                _testContext.SavedData["operationDuration"] = stopwatch.ElapsedMilliseconds;
                _testContext.LastResponse = response;
                _testContext.LastError = null;
            }
            catch (Exception ex)
            {
                stopwatch.Stop();
                _testContext.SavedData["operationDuration"] = stopwatch.ElapsedMilliseconds;
                _testContext.LastError = ex;
                _testContext.LastResponse = null;
            }
        }

        [Then(@"all tuples should be written successfully")]
        public void ThenAllTuplesShouldBeWrittenSuccessfully()
        {
            if (_testContext.LastError != null)
            {
                throw new Exception($"Expected successful response but got error: {_testContext.LastError.Message}");
            }

            // Validate that all tuples were written successfully
            // Implementation would check response status for each tuple
        }

        // ConfigureAwait and Async Patterns
        [Then(@"all requests should complete successfully")]
        public void ThenAllRequestsShouldCompleteSuccessfully()
        {
            if (!_testContext.SavedData.TryGetValue("concurrentResults", out var resultsObj))
            {
                throw new Exception("No concurrent results found");
            }

            // All requests completed successfully if we got here without exception
        }

        [Then(@"no race conditions should occur")]
        public void ThenNoRaceConditionsShouldOccur()
        {
            // This would typically involve checking for race conditions
            // In .NET, this might involve validating that shared state remains consistent
        }

        [Then(@"the client should remain thread-safe")]
        public void ThenTheClientShouldRemainThreadSafe()
        {
            // Validate that the client can handle concurrent access safely
            // This might involve checking internal client state or metrics
        }
    }
}
